DQC/ETL Layer Implementation

Dybal-Ilia opened the pull request:
# 1. Installation
The uv package manager is being used, here are steps to setup the environment:
## 1. Install uv package manager
- via *windows powerhell*:
```powershell -ExecutionPolicy ByPass -c "irm https://astral.sh/uv/install.ps1 | iex"``` for windows
- via *PyPi*:
- ```pipx install uv``` or ```pip install uv```
- via *curl*:
- ```curl -LsSf https://astral.sh/uv/install.sh | sh```
- via *wget*:
- ```wget -qO- https://astral.sh/uv/install.sh | sh```
## 2. Install dependencies
Dependencies and their versions are provided in *pyproject.toml* file
Anyways use:
 ```uv sync```

# 2. Implemented Features

## 1. DQC Layer
- Missing values detection
- Unique values detection
- Outliers detection (via IsolationForest)
- Duplicates detection
- Statistics(including user-provided procentiles)
- Inconsistancies detection (temporary negative values detection only)

## ETL Layer
- Data extraction (via opendatasets)
- Missing values handling (based on the provided strategy)
- Duplicates handling
- Negative values handling
- Processing categorical features (lowercasing values)
- Processing date
- Feature extension (experimental for feature creation)

## 3. Jupyter Notebook
- DQC implementation
- ETL implementation
- General data assumptions and idea explanations

Link to the task - https://miro.com/app/board/uXjVPBiA7pI=/?openCardPanel=3458764539070499580

Dybal-Ilia committed:
Initial commit

Commit:
DS-1.1 DQC and ETL

Commit:
Revert "DS-1.1 DQC and ETL"

This reverts commit 113b767e9c860c697d8bc0a00e338845a9279ce8.

Commit:
DS-1.1 DQC and ETL chanhges: -Added initial environment files

Commit:
DS-1.1 Added dataloader

Commit:
DS/1.1 Added dqc pipeline along with dqc template

Dybal-Ilia committed:
Delete mistakenly added separate dqc directory

Commit:
DS/1.1 Added decorators directory with exec_time decorator for time execution monitoring

Commit:
Merge branch 'DS/1.1' of https://github.com/Dybal-Ilia/Practice-DS-1.1 into DS/1.1

Commit:
DS/1.1 Updated dependencies

Commit:
DS-1.1 Changed QDC Layer logic

Commit:
DS-1.1 Updated outliers detection

Commit:
DS/1.1 Added dataloader

Commit:
DS/1.1 Added ETL Layer

Commit:
DS/1.1 Added initial functionality of Plotter class

Commit:
DS/1.1 Added merge_data function

Commit:
DS/1.1 Slightly changed DQCPieline behaviour

Commit:
DS/1.1 Slightly changed ETL layer behaviour

Commit:
DS/1.1 Slightly changed Plotter behaviour

Commit:
DS/1.1 Added notebooks directory for jupyter notebooks

Commit:
DS/1.1 Added logging and docs for methods

Commit:
DS/1.1 Updated dependencies

Commit:
DS/1.1 Added several explanations, fixed misspellings

Commit:
DS/1.1 Changed fatures to be trained on for outliers detection, fixed some misspellings

Commit:
DS/1.1 Slightly changed plotting parametres

Commit:
DS/1.1 Plots have been removed from dqc/etl and will be added as EDA layer

Commit:
DS/1.1 src/__init__.py file added

Commit:
DS/1.1 Updated dependencies

Aleksandra-Barysheva commented:
[Recommendations.txt](https://github.com/user-attachments/files/22022494/Dybal_Actionable_Recommendations.txt)

Commit:
DS/1.1 fix: outliers detection

Commit:
DS/1.1 feat: added datatypes analysis

added current type check

added check for mixed types

added check for numerics as strings

added check for possible datetime

added datetype suggestions

Commit:
DS/1.1 feat: added relationships analysis:

-check for left only

check for right only

-check for intersection

additionally methods names have been unified

Commit:
DS/1.1 feat: changed report rendering behaviour

-added separate method for validation report

-added separate method for statistics report

-added separate method for general report

Commit:
DS/1.1 feat: changed datatypes validation behaviour

- additionally added helper method for dtypes suggestion

Commit:
DS/1.1 feat: updated duplicates check behaviour

Commit:
DS/1.1 feat: changed statistics check behaviour

Commit:
DS/1.1 docs: added missing typehints and added docstrings for methods

Commit:
DS/1.1 refactor: methods code refactored so as to reduce complexity and make it more readable

Commit:
DS/1.1 fix: added self parameter where missing

Commit:
DS/1.1 feat: added apply methods:

-added apply_map for maps

-added apply_filter for filters

added helper methods for pipes application

Commit:
DS/1.1 redesign: reduntant methods were replaced with apply methods

Commit:
DS/1.1 refactor: merge_data method deleted

Commit:
DS/1.1 style: added typehints

Commit:
DS/1.1 fix: added missing duplicates report in validation_report

Commit:
DS/1.1 redesign: removed constructor, apply_map and apply_filter

Commit:
DS/1.1 redesign: dataloader realization moved to etl extract method, previous directory has been removed

Commit:
Ds/1.1 feat: added load method to conduct whole ETL pipeline

Commit:
DS/1.1 chore: added logging

Commit:
DS/1.1 feat: DQC and ETL emplemented

Commit:
DS/1.1 fix: fixed misusage of exec_time decorator

Commit:
DS/1.1 feat: added city extraction in pipeline
